### DISCUSSION TITLE
Use an external provider only for some file types

### DISCUSSION THREAD

**Discussion by renchap** [shrinerb/shrine#636] [Q&A]
I have an uploader that processes image & video files, with a somewhat complex configuration: derivatives, `determine_mime_type`, `store_dimensions`, several `add_metadata`…
Shrine is configured with backgrounding using Sidekiq:

```ruby
class ShrinePromoterWorker
  include Sidekiq::Worker
  sidekiq_options queue: 'shrine'

  def perform(attacher_class, record_class, record_id, name, file_data)
    attacher_class = Object.const_get(attacher_class)
    record = Object.const_get(record_class).find(record_id)

    attacher = attacher_class.retrieve(model: record, name:, file: file_data)

    attacher.file.open do
      attacher.refresh_metadata! if attacher.respond_to?(:refresh_metadata!)
      attacher.create_derivatives if attacher.respond_to?(:create_derivatives)
    end

    attacher.atomic_promote
  rescue Shrine::AttachmentChanged, ActiveRecord::RecordNotFound
    # attachment has changed or record has been deleted, nothing to do
  end
end

```

The derivatives & metadata extractions have 2 paths depending if the file is an image or a video.

As we get more and more videos, I would like to use an external service for video metadata extraction, derivatives generation and thumbnailing. https://www.coconut.co does this all at a cheap price, so this is my choice for now.

I am unsure on how to properly implement this. I have the code to send jobs to Coconut (giving them a list of formats I want, and signed URIs to upload the files to) and I can receive their callback on an URL once the job is done.

I store the mime type when creating the record, so this information is present before any processing (no need for `determine_mime_type`, this is currently only used as a security against spoofing).
So I need Shrine to somehow detect (before processing) that this file is a video, and then dont process it further but creates a Coconut job for it. Is this doable with Shrine? Should it be implemented in the `ShrinePromoteWorker`, for example `attacher.submit_to_coconut if attacher.use_coconut?`?

Then I guess I can have a custom controller getting the Coconut payload and updating the record directly, writing the correct fields? Or is it a better way?

**Comment by janko** [comment]
Hi @renchap, you probably already managed to write some integration code, but in case it helps, [shrine-transloadit](https://github.com/shrinerb/shrine-transloadit) is an example of using an external service to do the processing, then building uploaded file objects from the response. You can take a look at the source code to see how this is implemented.

**Reply by renchap** [reply]
I indeed wrote some code, but it is split between multiple files and there must be better ways of doing this.

I had a look at `shrine-transloadit`, but my main issue is that I want the same uploader to have 2 different ways of processing the files, depending on some conditions, which I did not manage to implement properly as a plugin.

I dont think my use-case is of interest to anybody else, so I abandoned the plugin idea :)
